prov/tcp/src/tcpx_conn_mgr.c

```diff
@@ -178,48 +180,152 @@ static int handle_poll_list(struct poll_fd_mgr *poll_mgr)
+			       struct tcpx_ep *ep,
+			       int index)
 {
-	struct tcpx_ep *ep;
-	struct fi_eq_cm_entry cm_entry;
-	struct fi_eq_err_entry err_entry;
+	struct tcpx_conn_req conn_req;
 	socklen_t len;
-	int ret, status;
+	int status, ret = 0;;
 
-	assert(poll_info->fid->fclass == FI_CLASS_EP);
-	ep = container_of(poll_info->fid, struct tcpx_ep, util_ep.ep_fid.fid);
+	assert(poll_mgr->poll_fds[index].revents == POLLOUT);
 
 	len = sizeof(status);
 	ret = getsockopt(ep->conn_fd, SOL_SOCKET, SO_ERROR, &status, &len);
 	if (ret < 0 || status) {
 		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "connection failure\n");
+		ret = status? status : errno;
+		goto err;
+	}
 
-		memset(&err_entry, 0, sizeof err_entry);
-		err_entry.fid = poll_info->fid;
-		err_entry.context = poll_info->fid->context;
-		err_entry.err = status ? status : errno;
+	conn_req.hdr.version = OFI_CTRL_VERSION;
+	conn_req.hdr.type = ofi_ctrl_connreq;
+	conn_req.cm_data_sz = htons(poll_info->cm_data_sz);
 
-		ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
-				  &err_entry, sizeof(err_entry), UTIL_FLAG_ERROR);
-	} else {
-		memset(&cm_entry, 0, sizeof cm_entry);
-		cm_entry.fid = poll_info->fid;
+	ret = tcpx_cm_send(ep->conn_fd, &conn_req, sizeof(conn_req));
+	if (ret)
+		goto err;
 
-		ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_CONNECTED,
-				  &cm_entry, sizeof(cm_entry), 0);
+	if (conn_req.cm_data_sz) {
+		ret = tcpx_cm_send(ep->conn_fd, poll_info->cm_data,
+				   poll_info->cm_data_sz);
+		if (ret)
+			goto err;
 	}
+err:
+	return ret;
+}
 
-	if (ret < 0)
+static int proc_recv_resp(struct poll_fd_mgr *poll_mgr,
+			  struct poll_fd_info *poll_info,
+			  struct tcpx_ep *ep,
+			  int index)
+{
+	struct tcpx_conn_req conn_resp;
+	struct fi_eq_cm_entry *cm_entry;
+	uint16_t cm_data_sz;
+	int ret = 0;
+
+	assert(poll_mgr->poll_fds[index].revents == POLLIN);
+	ret = tcpx_cm_recv(ep->conn_fd, &conn_resp, sizeof(conn_resp));
+	if (ret)
+		goto err;
+
+	if (conn_resp.hdr.type != ofi_ctrl_connresp) {
+		ret = -FI_ECONNREFUSED;
+		goto err;
+	}
+
+	if (conn_resp.hdr.version != OFI_CTRL_VERSION) {
+		ret = -FI_ENOPROTOOPT;
+		goto err;
+	}
+
+	cm_data_sz = ntohs(conn_resp.cm_data_sz);
+	if (cm_data_sz) {
+		if (cm_data_sz > TCPX_MAX_CM_DATA_SIZE)
+			goto err;
+
+		ret = tcpx_cm_recv(ep->conn_fd, poll_info->cm_data,
+				   cm_data_sz);
+		if (ret)
+			goto err;
+	}
+
+	cm_entry = calloc(1, sizeof(*cm_entry) + cm_data_sz);
+	if (!cm_entry) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "mem alloc failed\n");
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	cm_entry->fid = poll_info->fid;
+	memcpy(cm_entry->data, poll_info->cm_data, cm_data_sz);
+
+	ret = fi_eq_write(&ep->util_ep.eq->eq_fid, FI_CONNECTED,
+			  cm_entry, sizeof(*cm_entry) + cm_data_sz, 0);
+	if (ret < 0) {
 		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL, "Error writing to EQ\n");
+		goto err;
+	}
+
+	ret = fi_fd_nonblock(ep->conn_fd);
+err:
+	return ret;
+}
+
+static int handle_connect(struct poll_fd_mgr *poll_mgr,
+			  int index)
```

Comment:
The return values from this function aren't intuitive.  Most functions should be returning 0 on success and -errno on failure.